File: mobile/ios/packages/town-crier-data/Sources/API/APIEndpoint.swift
import Foundation

/// Describes an API endpoint with method, path, optional body, and query parameters.
public struct APIEndpoint: Sendable {
    public let method: HTTPMethod
    public let path: String
    public let body: (any Encodable & Sendable)?
    public let queryItems: [URLQueryItem]?

    public init(method: HTTPMethod, path: String, body: (any Encodable & Sendable)? = nil, queryItems: [URLQueryItem]? = nil) {
        self.method = method
        self.path = path
        self.body = body
        self.queryItems = queryItems
    }

    public static func get(_ path: String, query: [URLQueryItem]? = nil) -> APIEndpoint {
        APIEndpoint(method: .get, path: path, queryItems: query)
    }

    public static func post(_ path: String, body: some Encodable & Sendable) -> APIEndpoint {
        APIEndpoint(method: .post, path: path, body: body)
    }

    public static func put(_ path: String, body: some Encodable & Sendable) -> APIEndpoint {
        APIEndpoint(method: .put, path: path, body: body)
    }

    public static func delete(_ path: String) -> APIEndpoint {
        APIEndpoint(method: .delete, path: path)
    }

    public static func patch(_ path: String, body: some Encodable & Sendable) -> APIEndpoint {
        APIEndpoint(method: .patch, path: path, body: body)
    }
}

public enum HTTPMethod: String, Sendable {
    case get = "GET"
    case post = "POST"
    case put = "PUT"
    case delete = "DELETE"
    case patch = "PATCH"
}

File: mobile/android/data/src/main/kotlin/uk/towncrierapp/data/api/ApiEndpoint.kt
package uk.towncrierapp.data.api

/**
 * Describes a request `ApiClient` should make: method, path, an optional
 * already-JSON-encoded [body] (`null` means no request body at all — e.g.
 * `POST /v1/me` — not an empty string), and optional query parameters.
 * `internal`: only `:data`'s own repository implementations construct these
 * (port of iOS `APIEndpoint`).
 */
internal data class ApiEndpoint(
    val method: String,
    val path: String,
    val body: String? = null,
    val query: List<Pair<String, String>> = emptyList(),
) {
    companion object {
        fun get(
            path: String,
            query: List<Pair<String, String>> = emptyList(),
        ) = ApiEndpoint(method = "GET", path = path, query = query)

        fun post(
            path: String,
            body: String? = null,
        ) = ApiEndpoint(method = "POST", path = path, body = body)

        fun put(
            path: String,
            body: String? = null,
        ) = ApiEndpoint(method = "PUT", path = path, body = body)

        fun patch(
            path: String,
            body: String,
        ) = ApiEndpoint(method = "PATCH", path = path, body = body)

        fun delete(path: String) = ApiEndpoint(method = "DELETE", path = path)
    }
}
